langchain/embeddings/openai.py

```diff
@@ -473,9 +473,14 @@ def embed_documents(
         Returns:
             List of embeddings, one for each text.
         """
-        # NOTE: to keep things simple, we assume the list may contain texts longer
-        #       than the maximum context and use length-safe embedding function.
-        return self._get_len_safe_embeddings(texts, engine=self.deployment)
```

Comment:
get_len_safe_embeddings does batching under the hood, this would undo that

Reply:
Thanks for taking a look, @baskaryan. `_embedding_func` calls `_get_len_safe_embeddings` under the hood, so wouldn't the functionality remain the same for any text longer than `self.embedding_ctx_length`? https://github.com/hwchase17/langchain/blob/3e90c9798828c083445e7a57eee6d00256b9b569/langchain/embeddings/openai.py#L427-L443

Reply:
We need to be able to call the OpenAI endpoint by passing in text instead of tokens, and we are not currently able to do so with the `_get_len_safe_embeddings` function. Open to any alternative solutions.

Reply:
does embed_query or embed_documents not work?

Reply:
`embed_documents` is what I'm modifying here. `embed_query` will not work, since the vector stores are calling `embed_documents` under the hood: https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/chroma.py#L171

Reply:
yea but here it sends tokens (potentially across texts) up in batches

https://github.com/hwchase17/langchain/blob/master/langchain/embeddings/openai.py#L357-L363

out of curiosity, what's the use case?

Reply:
Right, but isn't that only necessary for longer pieces of text? The functionality for longer text would remain the same. The only change being made here is for shorter texts, which I don't suppose require batching anyway?

The use case is https://github.com/mlc-ai/mlc-llm/pull/551. We are enabling document QA for a wide variety of models that are run locally through an OpenAI compatible REST interface. We can't use the tokenized text, since the tokenizer assumes we're using the OpenAI models. Please let me know if I can provide any additional info!

Reply:
why wouldn't sending a batched call of many short texts be better than sending them one by one?

use case makes sense! but like you said _embedding_func can still call _get_len_safe_embeddings (unless you're setting embedding_ctx_length to be very high?) so does this fix?

what if we made using _get_len_safe_embeddings optional by making `embedding_ctx_length: Optional[int]` and if it's set to None then we change embed_documents to be something like

```python
        if self.embedding_cx_length:
          return self._get_len_safe_embeddings(texts, engine=self.deployment)
               _chunk_size = chunk_size or self.chunk_size

        if self.show_progress_bar:
            try:
                import tqdm

                _iter = tqdm.tqdm(range(0, len(documents), _chunk_size))
            except ImportError:
                _iter = range(0, len(documents), _chunk_size)
        else:
            _iter = range(0, len(documents), _chunk_size)

        for i in _iter:
            response = embed_with_retry(
                self,
                input=documents[i : i + _chunk_size],
                **self._invocation_params,
            )
            batched_embeddings += [r["embedding"] for r in response["data"]]

         for i in range(len(texts)):
            _result = results[i]
            if len(_result) == 0:
                average = embed_with_retry(
                    self,
                    input="",
                    **self._invocation_params,
                )[
                    "data"
                ][0]["embedding"]
            else:
                average = np.average(_result, axis=0, weights=num_tokens_in_batch[i])
            batched_embeddings[i] = (average / np.linalg.norm(average)).tolist()
```

basically implement a batched call without tokenizing

Reply:
@baskaryan that sounds good to me, but would it be okay to make `embedding_ctx_length` None by default? The Chroma vector store, which will be calling this function, won't let us pass in any additional arguments:

https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/vectorstores/chroma.py#L175

Reply:
That sounds good to me. I'll update the PR!